File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        int i=-1;
        for(auto& ch:s){
            if(ch=='(' || ch=='{' || ch=='[')
                s[++i]=ch;
            else{
                if(i>=0 && ((s[i]=='(' && ch==')') || (s[i]=='{' && ch=='}') || (s[i]=='[' && ch==']'))){
                    i--;
                }
                else{
                    return false;
                }
            }
        }
        return i==-1;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        char_map = {
            ")":"(",
            "]":"[",
            "}":"{"
        }

        stack = []
        for ch in s:
            if ch in char_map.values():
                stack.append(ch)
            elif ch in char_map.keys():
                if not stack or char_map[ch] != stack.pop():
                    return False
        
        return not stack
        